### PR TITLE
[13.0][IMP] shopfloor: zone picking

### PR DESCRIPTION
**Pull request by sebalix**
- [x] `start` -> `/scan_location` -> `start`
- [x] `start` -> `/scan_location` -> `select_picking_type`
---
- [x] `select_picking_type` -> `/list_move_lines` -> `select_line`
---
- [x] `select_line` -> `/list_move_lines` -> `select_line` (ordered by priority/location)
- [x] `select_line` -> `/scan_source` -> `select_line` (filtered on location)
- [x] `select_line` -> `/scan_source` -> `set_line_destination` (scanning a location/package/product/lot)
---
- [x] `set_line_destination` -> `/set_destination` -> `set_line_destination`
- [x] `set_line_destination` -> `/set_destination` -> `select_line`
- [x] <s>`set_line_destination` -> `/set_destination` -> `confirm_set_line_destination` (TODO to refactor to use a `confirmation` parameter on `/set_destination` as other scenarios?)</s>
- [x] <s>`confirm_set_line_destination` -> `/set_destination` -> `select_line`</s>
- [x] `set_line_destination` -> `/set_destination` -> `zero_check`
----
- [x] `zero_check` -> `/is_zero` -> `select_line` (**TODO :** still need to rewrite the inventory part to create the inventory on the location instead of location+product, both for zone & cluster picking scenario)
---
- [x] `stock_issue` -> `/stock_issue` -> `set_line_destination`
- [x] `stock_issue` -> `/stock_issue` -> `select_line`
---
=> Refactored to use the same code base with "Cluster Picking" scenario. NOTE: Almost done, usable, need more tests for corner cases.
- [x] `change_pack_lot` -> `/change_pack_lot` -> `change_pack_lot`
- [x] `change_pack_lot` -> `/change_pack_lot` -> `set_line_destination` 
- [x] `change_pack_lot` -> `/change_pack_lot` -> `select_line`
---
- [x] `select_line` -> `/prepare_unload` -> `select_line`
- [x] `select_line` -> `/prepare_unload` -> `unload_set_destination`
- [x] `select_line` -> `/prepare_unload` -> `unload_all`
---
- [x] `unload_all` -> `/set_destination_all` -> `unload_all`
- [x] `unload_all` -> `/set_destination_all` -> `select_line`
- [x] `unload_all` -> `/set_destination_all` -> `start`
- [x] `unload_all` -> `/unload_split` -> `unload_single`
- [x] `unload_all` -> `/unload_split` -> `unload_set_destination`
- [x] `unload_all` -> `/unload_split` -> `select_line`
---
- [x] `unload_single` -> `/unload_scan_pack` -> `unload_single`
- [x] `unload_single` -> `/unload_scan_pack` -> `unload_set_destination`
- [x] `unload_single` -> `/unload_scan_pack` -> `select_line`
- [x] `unload_single` -> `/unload_scan_pack` -> `start`
---
- [x] `unload_set_destination` -> `/unload_set_destination` -> `unload_single`
- [x] `unload_set_destination` -> `/unload_set_destination` -> `unload_set_destination` (+/- `confirmation_required`)
- [x] `unload_set_destination` -> `/unload_set_destination` -> `select_line`
- [x] `unload_set_destination` -> `/unload_set_destination` -> `start`

### CODE REVIEW

File: shopfloor/services/zone_picking.py [hunk 1]
```diff
@@ -0,0 +1,887 @@
...
+        * select_line: show the list of move lines
+        """
+        zone_location = self.env["stock.location"].browse(zone_location_id)
+        if not zone_location.exists():
+            return self._response_for_start(message=self.msg_store.record_not_found())
+        picking_type = self.env["stock.picking.type"].browse(picking_type_id)
+        if not picking_type.exists():
+            return self._response_for_start(message=self.msg_store.record_not_found())
+        move_lines = self._find_location_move_lines(zone_location, picking_type)
+        # TODO 'order' handling needs to be refactored
```
Comment (simahawk) [inline shopfloor/services/zone_picking.py]: @sebalix it might be probably better to not do this on server side. I can easily do it on client side. 
Do you have any specific requirements to keep this server side?

File: shopfloor/services/zone_picking.py [hunk 2]
```diff
@@ -0,0 +1,887 @@
...
+        if order == "location":
+            move_lines = move_lines.sorted(lambda l: l.location_id.name)
+        elif order == "priority":
+            move_lines = move_lines.sorted(lambda l: l.move_id.priority, reverse=True)
```
Comment (guewen) [inline shopfloor/services/zone_picking.py]: Should be extracted in a method for the sorting. The sorting by location sequence will need to be added (see #29)

Comment (sebalix) [reply shopfloor/services/zone_picking.py]: Added here: https://github.com/camptocamp/wms/pull/33/commits/2ffa9655ad6a8886eecd5961ebea077a50e691e0#diff-a87d3e49453610df30832e31c3d79f7eR224-R255

File: shopfloor/services/zone_picking.py (hunk 1 above)
Comment (sebalix) [reply shopfloor/services/zone_picking.py]: Checked on Slack, we preserve the order server-side.

**Comment by guewen** [comment]
@sebalix could you review https://github.com/camptocamp/wms/pull/34 and apply the same change here?

**Comment by sebalix** [comment]
@guewen I already done it here: https://github.com/camptocamp/wms/commit/2ffa9655ad6a8886eecd5961ebea077a50e691e0#diff-a87d3e49453610df30832e31c3d79f7eR247
But I could drop the `(lambda, reverse)` things to only use a list of fields to sort on as you did.

File: shopfloor/services/zone_picking.py [hunk 3]
```diff
@@ -280,7 +361,68 @@ def scan_source(self, zone_location_id, picking_type_id, barcode, order="priorit
...
+                )
+        # barcode not found, get back on 'select_line' screen
+        if not move_line:
+            response = self.list_move_lines(zone_location.id, picking_type.id)
+            return self._response(
+                base_response=response, message=self.msg_store.barcode_not_found(),
+            )
+        return self._response_for_set_line_destination(
+            zone_location, picking_type, move_line
+        )
```
Comment (guewen) [inline shopfloor/services/zone_picking.py:459]: This bulky one could be split in several methods :)

File: shopfloor/actions/operation.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+from odoo.tools.float_utils import float_compare
+
+from odoo.addons.component.core import Component
+
+
+class OperationAction(Component):
```
Comment (guewen) [inline shopfloor/actions/operation.py]: I'm not fond of "operation" which doesn't really mean anything (or mean too many things in fact).

Maybe these could be on models though? `split_othen_move_lines` on StockMove and `planned_qty_in_location_is_empty` on StockLocation?

Comment (sebalix) [reply shopfloor/actions/operation.py]: Updated, better like this indeed, it makes the code clear about the intent (especially the `move.split_other_move_lines(move_line)`).

File: shopfloor/services/zone_picking.py [hunk 5]
```diff
@@ -0,0 +1,1080 @@
...
+            next_state="confirm_unload_set_destination",
+            data=self._data_for_move_line(zone_location, picking_type, move_line),
+        )
+
+    def _data_for_select_picking_type(self, zone_location, picking_types):
+        return {
+            "zone_location": self.data.location(zone_location),
+            # available picking types to choose from
+            # TODO add lines count and priority lines count in the data
+            "picking_types": self.data.picking_types(picking_types),
```
Comment (simahawk) [inline shopfloor/services/zone_picking.py:186]: to agree on key names, please call the extra keys like this:

`lines_count`  # total lines count
`picking_count` # total operations/picking count
`priority_lines_count` # total lines count from priority pickings
`priority_picking_count` # total operations/picking count w/ high priority

**Comment by simahawk** [comment]
@sebalix can we drop the state `confirm_set_line_destination` (and other `confirm_*` states) and redirect to the same state + use `confirmation_required` flag as done per `location_content_transfer`?

File: shopfloor/services/zone_picking.py (hunk 5 above)
Comment (sebalix) [reply shopfloor/services/zone_picking.py:186]: Added in https://github.com/camptocamp/wms/pull/33/commits/2924da9cdee470188a0347bca1e96bc82b771c4b
Not sure about how to compute these numbers exactly, but a first implementation is done.

File: shopfloor/actions/data.py [hunk 6]
```diff
@@ -214,3 +214,54 @@ def picking_batches(self, record, with_pickings=False, **kw):
...
+
+    def picking_type(self, record, **kw):
+        parser = self._picking_type_parser
+        return self._jsonify(record, parser, **kw)
+
+    def picking_types(self, record, **kw):
+        return self.picking_type(record, multi=True)
+
+    @property
+    def _picking_type_parser(self):
```
Comment (simahawk) [inline shopfloor/actions/data.py:233]: @sebalix it's probably better to add computed fields for this or a single Serialized computed fields because you'll save some queries (you get 4 now).
Maybe just add a TODO.

File: shopfloor/actions/message.py [hunk 7]
```diff
@@ -155,6 +173,12 @@ def no_pack_in_location(self, location):
             "body": _("Location %s doesn't contain any package." % location.name),
         }
 
+    def several_lines_in_location(self, location):
+        return {
+            "message_type": "warning",
+            "body": _("Several lines found in %s, please scan one." % location.name),
```
Comment (simahawk) [inline shopfloor/actions/message.py]: translation broken: move out of () the interpolation.

File: shopfloor/models/stock_location.py [hunk 8]
```diff
@@ -35,3 +36,26 @@ def _get_reserved_move_lines(self):
...
+        have to look at the qty_done of the move lines from this location.
+        """
+        self.ensure_one()
+        quants = self.env["stock.quant"].search(
+            [("quantity", ">", 0), ("location_id", "=", self.id)]
+        )
+        remaining = sum(quants.mapped("quantity"))
+        lines = self.env["stock.move.line"].search(
+            [
+                ("state", "!=", "done"),
```
Comment (simahawk) [inline shopfloor/models/stock_location.py:57]: you should exclude "cancel / draft" no?

Comment (sebalix) [reply shopfloor/models/stock_location.py:57]: Probably, but as `state` is a related on `move_id.state`, and that we get no move lines on a draft/cancel move, it's maybe useless. But I agree it would be clearer by doing so.

**Comment by simahawk** [comment]
@sebalix I'm not sure this branch contains all latest changes from base one. Could you rebase if that's the case? Tnx!

**Comment by sebalix** [comment]
@simahawk squashed & rebased

**Comment by sebalix** [comment]
Squashed & rebased

**Comment by simahawk** [comment]
@sebalix the fixes done on the fly https://github.com/sebalix/wms/commit/1448936b56f2eda37897504c4d22ec04d36ee951
